feat: getting started carousel on dashboard page

dgolovin opened the pull request:
### What does this PR do?

(WIP) Creates Carousel component and adds Learning Center section to Dashboard page

### Screenshot / video of UI

Behavior example with the current list of guides

https://github.com/containers/podman-desktop/assets/620330/3b84156a-400c-4d30-850f-246d5a70861e

### What issues does this PR fix or reference?

Fix #4582.

<!-- Include any related issues from Podman Desktop 
repository (or from another issue tracker). -->

### How to test this PR?

<!-- Please explain steps to reproduce -->

benoitf commented on packages/main/src/plugin/learning-center/learning-center.ts:
```diff
@@ -0,0 +1,26 @@
+ ***********************************************************************/
+
+import type { Guide } from './learning-center-api.js';
+import { default as guidesJson } from './guides.json';
```
it's something new not used inside Podman Desktop but now you can use typescript import attributes

```suggestion
import { default as guidesJson } from './guides.json' with { type: "json" };
```

dgolovin replied:
`Import attributes are only supported when the '--module' option is set to 'esnext' or 'nodenext'.`

benoitf commented on packages/main/src/plugin/learning-center/learning-center.ts:
```diff
@@ -0,0 +1,26 @@
+import { default as guidesJson } from './guides.json';
+
+export class LearningCenter {
+  static get guides() {
+    return guidesJson.guides as Guide[];
+  }
```
not sure why we need a class if you use only static stuff inside

dgolovin replied:
Just to avoid undefined check, but if I use `with {type: "json"}` it would not be needed.

dgolovin commented:
I've updated demo to reflect latest changes.

cdrage reviewed (changes requested):
I'm finding the logo's to be absolutely massive. Any way we can limit the size? For example "Lima" is huge compared to the other provider logos.

cdrage commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,23 @@
+<div class="flex flex-1 flex-col bg-charcoal-600 pb-4 rounded-lg hover:bg-zinc-700">
+  <dif class="pt-4">
+    <div class="px-4">
+      <img src="{`data:image/png;base64,${guide.icon}`}" alt="{guide.id}" style="{`height:48px;`}" />
```
With height 48px the Lima logo spans the entire section, but the other logos look okay (docker, podman, etc.).

dgolovin replied:
That is the only way IMO to show logos that is not square, like Kind, Lima, NodeJS and etc.

dgolovin replied:
Updated guides.json and this is moot for now.

benoitf commented:
hi @dgolovin I'm not sure about the links being used.

I thought it should be links to tutorials like https://developers.redhat.com/articles/2023/09/05/beginners-guide-python-containers

not links to compose, kind, etc.

dgolovin commented:
@benoitf I'll update guides.json with links to developers.redhat.com related to podman.

slemeur commented on packages/main/src/plugin/learning-center/guides.json:
```diff
@@ -0,0 +1,44 @@
+      "id": "podman-desktop-for-kubernetes",
+      "url": "https://developers.redhat.com/articles/2023/11/06/working-kubernetes-podman-desktop",
+      "title": "Working with Kubernetes",
+      "description": "Understand how Podman's design philosophy can handle pods just as it handles containers, in a way that is consistent with Kubernetes architectures.",
```
```suggestion
      "description": "Discover the Kubernetes features in Podman, including Kubernetes integration, setting up a local environment using Kind or minikube, and interacting with applications.",
```

slemeur commented on packages/main/src/plugin/learning-center/guides.json:
```diff
@@ -0,0 +1,44 @@
+      "icon": "iVBORw0KGgoAAAANSUhEUgAAAfkAAAH5CAYAAACCtkfkAAABg2lDQ1BJQ0MgcHJvZmlsZQAAKJF9kT1Iw0AcxV9TtSIVh3YQdchQneyiIo5ahSJUCLVCqw4ml35Bk4YkxcVRcC04+LFYdXBx1tXBVRAEP0BcXZwUXaTE/yWFFjEeHPfj3b3H3TtAaFSYZnXNAppum+lkQszmVsXQK8KIoAfDCMnMMuYkKQXf8XWPAF/v4jzL/9yfo1/NWwwIiMSzzDBt4g3i6U3b4LxPHGUlWSU+Jx436YLEj1xXPH7jXHRZ4JlRM5OeJ44Si8UOVjqYlUyNeIo4pmo65QtZj1XOW5y1So217slfGM7rK8tcpzmCJBaxBAkiFNRQRgU24rTq ...
+    },
+    {
+      "id": "spring",
```
this looks like a duplicate of the previous one

slemeur commented on packages/main/src/plugin/learning-center/guides.json:
```diff
@@ -0,0 +1,44 @@
+    },
+    {
+      "id": "spring",
+      "url": "https://developers.redhat.com/articles/2023/07/03/managing-java-containers-quarkus-and-podman-desktop",
```
shouldn't it be https://developers.redhat.com/articles/2023/10/19/containerize-spring-boot-application-podman-desktop

lstocchi commented on packages/renderer/src/lib/carousel/Carousel.spec.ts:
```diff
@@ -0,0 +1,129 @@
+test('carousel cards get visible when size permits', async () => {
+  render(CarouselTest);
+  const card1 = screen.getByText('card 1');
+  console.log(window.innerWidth);
```
to be removed

dgolovin replied:
```suggestion
```

lstocchi commented on packages/renderer/src/lib/carousel/Carousel.spec.ts:
```diff
@@ -0,0 +1,129 @@
+test('carousel left and right buttons enabled when all items does not fit into screen and disabled otherwise', async () => {
+  render(CarouselTest);
+  const card1 = screen.getByText('card 1');
+  console.log(window.innerWidth);
```
to be removed

dgolovin replied:
```suggestion
```

lstocchi commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,23 @@
+</script>
+
+<div class="flex flex-1 flex-col bg-charcoal-600 pb-4 rounded-lg hover:bg-zinc-700">
+  <dif class="pt-4">
```
```suggestion
  <div class="pt-4">
```

lstocchi commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,23 @@
+<div class="flex flex-1 flex-col bg-charcoal-600 pb-4 rounded-lg hover:bg-zinc-700">
+  <dif class="pt-4">
+    <div class="px-4">
+      <img src="{`data:image/png;base64,${guide.icon}`}" alt="{guide.id}" style="{`height:48px;`}" />
```
```suggestion
      <img src="{`data:image/png;base64,${guide.icon}`}" alt="{guide.id}" style="height:48px;" />
```

lstocchi commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,23 @@
+      {guide.title}
+    </div>
+    <p class="text-base text-gray-700 px-4 pt-4">{guide.description}</p>
+  </dif>
```
```suggestion
  </div>
```

lstocchi reviewed (commented):
Looks fine to me. You just need to fix the duplicate error mentioned by Stevan 👍

dgolovin commented on packages/main/src/plugin/learning-center/guides.json:
```diff
@@ -0,0 +1,44 @@
+      "id": "spring",
+      "url": "https://developers.redhat.com/articles/2023/07/03/managing-java-containers-quarkus-and-podman-desktop",
+      "title": "Managing Java containers with Quarkus",
+      "description": "Learn new skills to manage Java containers with Quarkus and Podman Desktop.",
```
```suggestion
      "description": "Learn how to take a Spring application, containerize it, and run it locally.",
```

dgolovin commented on packages/main/src/plugin/learning-center/guides.json:
```diff
@@ -0,0 +1,44 @@
+    {
+      "id": "spring",
+      "url": "https://developers.redhat.com/articles/2023/07/03/managing-java-containers-quarkus-and-podman-desktop",
+      "title": "Managing Java containers with Quarkus",
```
```suggestion
      "title": "Containerize a Spring Boot application with Podman Desktop",
```

lstocchi reviewed (commented):
This is something that @deboer-tim would comment 😆 but he's on PTO so i'll take his place.
Can we handle the card height differently? 
In the example below you can see that if i shrink podman desktop and start looping on the carousel, the arrows changes position. It would be great if they stay fixed. Maybe calculating a min-height or max-height and then cut the content by adding ... or something else.
 
![min-height-issue](https://github.com/containers/podman-desktop/assets/49404737/74e41a86-d7f2-4f8b-984b-70a9086cdf4a)

deboer-tim commented:
> This is something that @deboer-tim would comment 😆 but he's on PTO so i'll take his place.

+1 to comment, constant height is exactly what I would say if I was here. 😉😛

dgolovin commented:
There is no perfect solution for this. Imagine we have fixed it and the h is the same. Then we have an opposite problem for this state

![image](https://github.com/containers/podman-desktop/assets/620330/fe78cabf-047b-4956-8a86-339813d0406c)

IMO it is the same level of 'bad', but it comes with additional code we don't have right now. Of course it is possible to adjust description and make all the boxes to be filled evenly, but the came can be done for current implementation so h would jump less.

deboer-tim commented:
> There is no perfect solution for this. Imagine we have fixed it and the h is the same. Then we have an opposite problem for this state
> 
> IMO it is the same level of 'bad', but it comes with additional code we don't have right now. Of course it is possible to adjust description and make all the boxes to be filled evenly, but the came can be done for current implementation so h would jump less.

Honestly, I think this is 'better', but the real answer is that we should do both. Having a consistent height is good, but it should be ~25% smaller (i.e. not sized for the biggest title/description that could ever exist), and we should edit cards (probably just make Kubernetes description shorter, remove redundant 'with Podman Desktop' from Spring title) so that the lengths are shorter and more consistent.

dgolovin commented:
@deboer-tim @lstocchi @cdrage @slemeur I've updated cards behavior, guides description and updated demo in pr description, please review when you have time.

lstocchi commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,26 @@
+
+<div
+  class="flex flex-col flex-1 bg-charcoal-600 pb-4 rounded-lg hover:bg-zinc-700 min-w-[{width}px] min-h-[{height}px]">
+  <dif class="pt-4 flex flex-col">
```
```suggestion
  <div class="pt-4 flex flex-col">
```

lstocchi commented on packages/renderer/src/lib/learning-center/GuideCard.svelte:
```diff
@@ -0,0 +1,26 @@
+      {guide.title}
+    </div>
+    <p class="line-clamp-4 px-4 pt-4">{guide.description}</p>
+  </dif>
```
```suggestion
  </div>
```

lstocchi reviewed (commented):
Tested and works fine for me. Just a wrong tag

lstocchi reviewed (approved):
Forgot to approve

deboer-tim reviewed (approved):
Cards fit in well, and scale nicely as you change the window. Thanks for all the work making this look really clean.

One nit: the descriptions are bright white and stand out too much compared with previous edits - could you please drop these back down to a grey before merging?

Personally I find the font a little large too vs other things on the dashboard, but I'm ok trying this in context for a while.

dgolovin commented:
@deboer-tim I've lost correct colors along the way. Thank you for pointing that out. Fixed. Now card use styles similar to other card on dashboard.

![image](https://github.com/containers/podman-desktop/assets/620330/2fe10200-1e81-4d27-9857-68cfd46e4f21)

benoitf reviewed (commented):
I see no tests for GuideCard.svelte and LearningCenter.svelte

dgolovin commented:
Created https://github.com/containers/podman-desktop/issues/6036.

benoitf commented:
thanks @dgolovin 👏
